modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/TransactionService.scala

```diff
@@ -3,37 +3,57 @@ package org.tessellation.dag.l1.domain.transaction
 import cats.data.NonEmptyList
 import cats.data.Validated.{Invalid, Valid}
 import cats.effect.Async
-import cats.syntax.applicative._
-import cats.syntax.either._
-import cats.syntax.flatMap._
-import cats.syntax.functor._
+import cats.syntax.all._
 
-import org.tessellation.node.shared.domain.transaction.ContextualTransactionValidator
-import org.tessellation.node.shared.domain.transaction.ContextualTransactionValidator.ContextualTransactionValidationError
+import org.tessellation.node.shared.domain.collateral.LatestBalances
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.schema.snapshot.{Snapshot, SnapshotInfo, StateProof}
 import org.tessellation.schema.transaction.Transaction
 import org.tessellation.security.Hashed
 import org.tessellation.security.hash.Hash
 
+import fs2.Stream
+
+import ContextualTransactionValidator.ContextualTransactionValidationError
+
 trait TransactionService[F[_]] {
   def offer(transaction: Hashed[Transaction]): F[Either[NonEmptyList[ContextualTransactionValidationError], Hash]]
 }
 
 object TransactionService {
 
-  def make[F[_]: Async](
+  def make[
+    F[_]: Async,
+    P <: StateProof,
+    S <: Snapshot,
+    SI <: SnapshotInfo[P]
+  ](
     transactionStorage: TransactionStorage[F],
+    lastSnapshotStorage: LastSnapshotStorage[F, S, SI] with LatestBalances[F],
     contextualTransactionValidator: ContextualTransactionValidator[F]
   ): TransactionService[F] = new TransactionService[F] {
 
     def offer(
       transaction: Hashed[Transaction]
     ): F[Either[NonEmptyList[ContextualTransactionValidationError], Hash]] =
-      contextualTransactionValidator.validate(transaction.signed).flatMap {
-        case Valid(_) =>
-          transactionStorage
-            .put(transaction)
-            .as(transaction.hash.asRight[NonEmptyList[ContextualTransactionValidationError]])
-        case Invalid(e) => e.toNonEmptyList.asLeft[Hash].pure[F]
+      lastSnapshotStorage.getOrdinalStream.zipWithPrevious.mapFilter {
```

Comment:
I think there is a nice operator for this task: `changes` that takes the `Eq` for compared element and only emits new elements.

Reply:
:+1: